alchemist-incarnation-collektive/src/main/kotlin/it/unibo/alchemist/CollektiveIncarnation.kt

```diff
@@ -58,39 +84,77 @@ class CollektiveIncarnation<P> : Incarnation<Any?, P> where P : Position<P> {
 
     override fun createReaction(
         randomGenerator: RandomGenerator,
         environment: Environment<Any?, P>,
         node: Node<Any?>,
         timeDistribution: TimeDistribution<Any?>,
         parameter: String,
-    ): Reaction<Any?> = Event(node, timeDistribution).also {
-        it.actions = ListSet.of(
-            createAction(randomGenerator, environment, node, timeDistribution, it, parameter),
-        )
-    }
+    ): Reaction<Any?> =
+        Event(node, timeDistribution).also {
+            it.actions =
+                ListSet.of(
+                    createAction(randomGenerator, environment, node, timeDistribution, it, parameter),
+                )
+        }
 
     override fun createTimeDistribution(
         randomGenerator: RandomGenerator,
         environment: Environment<Any?, P>,
         node: Node<Any?>?,
         parameter: String?,
-    ): TimeDistribution<Any?> = parameter?.toDoubleOrNull().let { frequency ->
-        val actualFrequency = frequency ?: 1.0
-        DiracComb(DoubleTime(randomGenerator.nextDouble(0.0, 1.0 / actualFrequency)), actualFrequency)
-    }
+    ): TimeDistribution<Any?> =
+        parameter?.toDoubleOrNull().let { frequency ->
+            val actualFrequency = frequency ?: 1.0
+            DiracComb(DoubleTime(randomGenerator.nextDouble(0.0, 1.0 / actualFrequency)), actualFrequency)
+        }
 
     override fun createNode(
         randomGenerator: RandomGenerator,
         environment: Environment<Any?, P>,
         parameter: String?,
-    ): Node<Any?> = GenericNode(environment).also {
-        it.addProperty(CollektiveDevice(environment, it, DoubleTime(parameter?.toDoubleOrNull() ?: 0.0)))
+    ): Node<Any?> =
+        GenericNode(environment).also {
+            it.addProperty(
+                CollektiveDevice(
+                    environment,
+                    it,
+                    parameter?.toDoubleOrNull().let { p ->
+                        if (p != null) DoubleTime(p) else null
+                    },
+                ),
+            )
+        }
+
+    companion object {
+        // a delegate
+        private object ScriptEngine {
+            operator fun getValue(thisRef: Any?, property: KProperty<*>) =
+                ScriptEngineManager().getEngineByName(property.name)
+                    ?: error("No script engine with ${property.name} found.")
+        }
+        private val kotlin by ScriptEngine
+        private val defaultLambda: (Any?) -> Any? = { Double.NaN }
+
+        private val cache: LoadingCache<String, (Any?) -> Any?> = Caffeine.newBuilder()
+            .build { property ->
+                runCatching {
+                    @Suppress("UNCHECKED_CAST")
+                    when (val interpreted = kotlin.eval(property)) {
+                        is (Nothing) -> Any? -> interpreted
+                        else -> defaultLambda
+                    } as (Any?) -> Any?
+                }.getOrElse { defaultLambda }
+            }
```

Comment:
What the fuck is going on here 🤣 
@angelacorte could you explain?

Reply:
@nicolasfara resolve if you are satisfied with the answer